core/interfaces/rebase.py

```diff
@@ -284,6 +296,95 @@ def base_ref(self):
+
+        relatives = []
+        for rel in branch_tree:
+            match = re.search(self.nearest_node_pattern, rel)
+            # print(match.groups(), branch, default)
+            if not match:
+                # util.debug.add_to_log('Could not parse relative string %s' % rel)
+                continue
+            branch_name = match.groups()[0]
+            if branch_name != branch and branch_name not in relatives:
+                relatives.append(branch_name)
+
+        util.debug.add_to_log('nearest_branch: found %s relatives: %s' %
+                              (len(relatives), relatives))
+
+        if not relatives:
+            util.debug.add_to_log('No relatives found. Possibly on a root branch!')
+            return default
+
+        util.debug.add_to_log('nearest_branch: filter out branches that do not'
+                              ' share branch-out nodes')
+        diff = difflib.Differ()
+        branch_commits = self.git("rev-list", "--first-parent", branch).splitlines()
+        nearest = None
+        max_iterations = 100
+        for relative in relatives:
+            util.debug.add_to_log('nearest_branch: Getting common commits with %s' % relative)
+            relative_commits = self.git("rev-list", "--first-parent", relative).splitlines()
+            common = None
+            for i, l in enumerate(diff.compare(branch_commits, relative_commits)):
+                if i >= max_iterations:
+                    util.debug.add_to_log('nearest_branch: Max iterations exceeded!')
+                    break
+                if not l.startswith(' '):
+                    util.debug.add_to_log('nearest_branch: commit differs %s' % l)
+                    continue
+                common = l.strip()
+                util.debug.add_to_log('nearest_branch: found common commit %s' % common)
+                break
+
+            if not common:
+                util.debug.add_to_log('nearest_branch: No common commit found with %s' % relative)
+                continue
+
+            # found common "branch-out" node with relative, get branches
+            branches = self.git("branch", "--contains", common,
+                                "--no-merged").splitlines()
+            cleaned_branch_names = [b[2:].strip() for b in branches]
+            util.debug.add_to_log('nearest_branch: got valid branches %s' %
+                                  cleaned_branch_names)
+            if relative in cleaned_branch_names:
+                nearest = relative
+                break
+
+        if not nearest:
+            util.debug.add_to_log('No valid nearest found. Possibly on a root '
+                                  '/ detached branch!')
+            return default
+
+        util.debug.add_to_log('nearest_branch: found best candidate %s' % nearest)
+
+        # if same as branch, return default instead
+        if branch == nearest:
+            return default
+        return nearest
+
```

Comment:
This is a huge function.  Would be great if it could be decomposed somehow, but its okay if that would add extra complexity or damage readability.